### PR TITLE
An article on source code analysis of kubeedge's overall architecture

### PR DESCRIPTION
**Pull request by WormOn**
* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:

### CODE REVIEW

**Comment by fisherxu** [comment]
Thanks! @WormOn 
/cc @kevin-wangzefeng @kadisi

File: content/zh/blog/source-code-analysis/kubeedge源码分析之整体架构.md [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+|:---:|:---:|:---:|
+|cloudcore|kubeedge/cloud|kubeedge/cloud/cloudcore/cloudcore.go，kubeedge/cloud/admission/admission.go，kubeedge/cloud/csidriver/csidriver.go|
+|edgecore|kubeedge/edge|kubeedge/edge/cmd/edgecore/edgecore.go|
+|edge_mesh|kubeedge/edgemesh|kubeedge/edgemesh/cmd/edgemesh.go|
+|edge_site|kubeedge/edgesite|kubeedge/edgesite/cmd/edgesite.go|
+
+在cloudcore、edgecore、edge_mesh和edge_site组件的源码中都使用了命令行框架[cobra](https://github.com/spf13/cobra) ,具体如下：
+
+1. cloudcore代码入口
+
```
Comment (kadisi) [inline content/zh/blog/source-code-analysis/kubeedge源码分析之整体架构.md]: please use  markdown code block. 
image

File: content/zh/blog/source-code-analysis/kubeedge源码分析之整体架构.md [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+	kubeedge/cloud/cloudcore/cloudcore.go
+	
+		func main() {
+			command := app.NewCloudCoreCommand() //此函数是对cobra调用的封装
+			...
+		}
+	
+	进入app.NewCloudCoreCommand()函数内部，也就是kubeedge/cloud/cloudcore/app/server.go中的NewCloudCoreCommand()函数中，具体如下：
+	
+		func NewCloudCoreCommand() *cobra.Command {
```
Comment (kadisi) [inline content/zh/blog/source-code-analysis/kubeedge源码分析之整体架构.md]: code block

File: content/zh/blog/source-code-analysis/kubeedge源码分析之整体架构.md [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+		}
+ 	在NewCloudCoreCommand()函数中，通过	registerModules()函数注册cloudcore中的功能模块，通过core.Run()函数启动已注册的cloudcore中的功能模块，至于registerModules()函数注册了哪些功能模块，core.Run()函数怎么启动已注册功能模块的，详见“组件中模块的共用框架和功能”。
+ 	
+> 注意：kubeedge/cloud/admission/admission.go，kubeedge/cloud/csidriver/csidriver.go两个入口，目前貌似还没有用到，暂不分析。
+
+ 2. edgecore代码入口
+
+	kubeedge/edge/cmd/edgecore/edgecore.go
+	
+		func main() {
```
Comment (kadisi) [inline content/zh/blog/source-code-analysis/kubeedge源码分析之整体架构.md]: code block

File: content/zh/blog/source-code-analysis/kubeedge源码分析之整体架构.md [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+			command := app.NewEdgeCoreCommand()//此函数是对cobra调用的封装
+			...
+		}
+进入app.NewEdgeCoreCommand()函数内部，也就是kubeedge/edge/cmd/edgecore/app/server.go中的NewEdgeCoreCommand()函数中，具体如下：
+
+		func NewEdgeCoreCommand() *cobra.Command {
```
Comment (kadisi) [inline content/zh/blog/source-code-analysis/kubeedge源码分析之整体架构.md]: code block

File: content/zh/blog/source-code-analysis/kubeedge源码分析之整体架构.md [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+			},
+		  }
+		  ...
+		}
+		
+	在NewEdgeCoreCommand()函数中，通过	registerModules()函数注册edgecore中的功能模块，通过core.Run()函数启动已注册的edgecore中的功能模块，至于registerModules()函数注册了哪些功能模块，core.Run()函数怎么启动已注册功能模块的，详见“组件中模块的共用框架和功能”。
+	
+3. edge_mesh代码入口
+
+	kubeedge/edgemesh/cmd/edgemesh.go
```
Comment (kadisi) [inline content/zh/blog/source-code-analysis/kubeedge源码分析之整体架构.md]: code block

File: content/zh/blog/source-code-analysis/kubeedge源码分析之整体架构.md [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+			//Start server
+			server.StartTCP() //启动一个tcp服务
+		}
+	
+	从main()函数中可以看到，edgemesh没有使用cobra，而是直接注册功能模块，然后启动了一个TCP服务。
+	
+4. edge_site代码入口
+
+	kubeedge/edgesite/cmd/edgesite.go
+
```
Comment (kadisi) [inline content/zh/blog/source-code-analysis/kubeedge源码分析之整体架构.md]: code block

File: content/zh/blog/source-code-analysis/kubeedge源码分析之整体架构.md [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+kubeedge组件中各个功能模块之间是通过Beehive来组织和管理的，Beehive是一个基于go-channels的消息框架，但本文的重点不是不是Beehive，所以只会分析kubeedge中用到的Beehive的相关功能。下面来深入cloudcore、edgecore、edge_mesh和edge_site组件中，一起探究组件内部各功能模块的共用框架。
+
+
+### cloudcore中模块的共用框架和功能分析
+
+在“组件的共用框架和功能”的“cloudcore代码入口”部分已经分析到cloudcore中功能模块的注册和已注册功能模块的启动，本节就接着往下分析。
+
+1. cloudcore中功能模块的注册
+
```
Comment (kadisi) [inline content/zh/blog/source-code-analysis/kubeedge源码分析之整体架构.md]: code block

File: content/zh/blog/source-code-analysis/kubeedge源码分析之整体架构.md (hunk 1 above)
Comment (WormOn) [reply content/zh/blog/source-code-analysis/kubeedge源码分析之整体架构.md]: 👌

**Pull request closed without merging**